library/alloc/src/string.rs

```diff
@@ -117,27 +117,84 @@ use crate::vec::Vec;
 ///
 /// # UTF-8
 ///
-/// `String`s are always valid UTF-8. This has a few implications, the first of
-/// which is that if you need a non-UTF-8 string, consider [`OsString`]. It is
-/// similar, but without the UTF-8 constraint. The second implication is that
-/// you cannot index into a `String`:
+/// `String`s are always valid UTF-8. If you need a non-UTF-8 string, consider
+/// [`OsString`]. It is similar, but without the UTF-8 constraint. Because UTF-8
+/// is a variable width encoding, `String`s can be smaller than an array of
+/// the same `chars`:
+///
+/// ```
+/// use std::mem;
+///
+/// // `s` is ASCII which represents each `char` as one byte
+/// let s = "hello";
+/// assert_eq!(s.len(), 5);
+///
+/// // A `char` array with the same contents would be longer because
+/// // every `char` is four bytes
+/// let s = ['h', 'e', 'l', 'l', 'o'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
+/// assert_eq!(size, 20);
+///
+/// // However, for non-ASCII strings, the difference will be smaller
+/// // and sometimes they are the same
+/// let s = "💖💖💖💖💖";
+/// assert_eq!(s.len(), 20);
+///
+/// let s = ['💖', '💖', '💖', '💖', '💖'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
+/// assert_eq!(size, 20);
+/// ```
+///
+/// This raises an interesting question as to how `s[i]` should work.
+/// What should `i` be here - a byte index or a `char` index? Because
+/// of UTF-8 encoding we can't find the `i`th `char` in constant-time.
+/// So it should be a byte index. But then what should `s[i]` return?
+/// It could be a `char` but that could be wasteful if the character
+/// starting at byte index `i` is encoded in UTF-8 with fewer than four bytes.
```

Comment:
`s[i]` can't actually return a char, since the Index trait [requires `&Self::Output`](https://doc.rust-lang.org/nightly/std/ops/trait.Index.html#tymethod.index). `s[i]` is `*s.index(i)` under the hood, but there's no way for String to return a `&char` in that method, since the char doesn't live anywhere in memory.

Reply:
That's a really good point I hadn't considered at all! Okay so `s[i]` could return `&char` if `String` was internally `Vec<char>` instead of `Vec<u8>`. We have `Vec<u8>` because of the UTF-8 encoding (because `Vec<char>` would be UTF-32?). So I guess the logic here is:
1. we could've chosen any encoding and we chose UTF-8
2. because we chose UTF-8 we store data as `Vec<u8>`
3. because we store data as `Vec<u8>`, `s.index(i)` can _only_ return a `&u8`

Is that correct? I can't tell if the 2nd option is logically required or if there are other ways of storing UTF-8 data

Reply:
In general 1+2 are correct (there's a 0. which is that UTF-8 is generally a pretty good default due to interoperability with ASCII and common usage throughout e.g. web etc).

We could return anything that's in the vec - for example, &[u8] or &str - but those are weird to return from a particular index accessed.

Reply:
Hopefully handled in a52ab6946e508fdfe4f8dc6313f9b7d9e0d77b58